pennylane/transforms/metric_tensor.py

```diff
@@ -14,105 +14,139 @@
 """
 Contains the metric tensor transform
 """
+import functools
 import warnings
 
 import numpy as np
 import pennylane as qml
 
 
+from .batch_transform import batch_transform
+
+
+SUPPORTED_OPS = ["RX", "RY", "RZ", "PhaseShift"]
+
+
 def _stopping_critera(obj):
-    if getattr(obj, "num_params", 0) == 0:
-        return True
+    return getattr(obj, "num_params", 0) == 0 or obj.name in SUPPORTED_OPS
 
-    if obj.name in ["RX", "RY", "RZ", "PhaseShift"]:
-        return True
 
-    return False
+def expand_fn(tape):
+    """Expands the tape to contain only operations
+    supported by the ``metric_tensor`` transform (specified
+    by ``SUPPORTED_OPS``).
+    """
+    new_tape = tape.expand(depth=2, stop_at=_stopping_critera)
+    params = new_tape.get_parameters(trainable_only=False)
+    new_tape.trainable_params = qml.math.get_trainable_indices(params)
+    return new_tape
+
 
+@functools.partial(batch_transform, expand_fn=expand_fn)
+def metric_tensor(tape, diag_approx=False):
+    """Returns a function that computes the block-diagonal approximation of the metric tensor
+    of a given QNode or quantum tape.
 
-def metric_tensor_tape(tape, diag_approx=False, wrt=None):
-    """Returns a list of tapes, and a classical processing function, for computing the block
-    diagonal metric tensor approximation of an input tape on hardware.
+    .. note::
+
+        Currently, only the :class:`~.RX`, :class:`~.RY`, :class:`~.RZ`, and
+        :class:`~.PhaseShift` parametrized gates are supported.
+        All other parametrized gates will be decomposed if possible.
 
     Args:
-        tape (.QuantumTape): the tape to compute the metric tensor of
-        diag_approx (bool): If ``True`` the diagonal approximation to the metric
-            tensor is computed. If ``False``, a block diagonal approximation
-            to the metric tensor is computed.
-        wrt (Sequence[int]): Indices of the tape parameters with which to
-            compute the metric tensor. Parameter indices not included are
-            treated as *fixed* parameters. Defaults to the tape's trainable
-            parameters.
+        qnode (.QNode or .QuantumTape): quantum tape or QNode to find the metric tensor of
+        diag_approx (bool): if True, use the diagonal approximation

```

Comment:
```suggestion
        diag_approx (bool): if True, use the diagonal approximation. If ``False``, a 
        block diagonal approximation of the metric tensor is computed.
```

Reply:
Took me a second to realise there is a distinction between diagonal approx and block diagonal approx here

Reply:
A majority of this docstring was written _before_ the first QNG paper came out, so the 'default' assumption was that the metric tensor could only be block diagonal -- we didn't know how at the time to get the full metric tensor on hardware!

However, since then, there have been papers showing how to get the full metric tensor:

- [Using the linear combination of unitaries/Hadamard test approach](https://arxiv.org/abs/1901.00015)
- [The parameter-shift rule](https://arxiv.org/abs/2008.06517)
- [The adjoint method](https://arxiv.org/abs/2011.02991)

The first two work with hardware, the latter is simulation only. We should add them to PL :)